File: neorg_lsp/rpc/initialize.py
from typing import Any

import attrs

from neorg_lsp.rpc import BaseRequest, BaseResponse


@attrs.define
class InitializeRequestParams:
    capabilities: dict[Any, Any] = dict()
    processId: int | None = None
    rootPath: str | None = None


@attrs.define
class InitializeRequest(BaseRequest):
    params: InitializeRequestParams = InitializeRequestParams()

@attrs.define
class CompletionOptions:
    resolveProvider: bool
    triggerCharacters: list[str]

@attrs.define
class ServerCapabilities:
    completionProvider: CompletionOptions


@attrs.define
class InitializeParams:
    capabilities: ServerCapabilities


@attrs.define
class InitializeResult(BaseResponse):
    result: InitializeParams | None = None

    @staticmethod
    def success(id: int, result: InitializeParams):
        req = InitializeResult(id)
        req.result = result
        return req

File: CeorgLsp/Rpc/Initialize.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace CeorgLsp.Rpc
{
    public record InitializeRequest : Request
    {
        [JsonPropertyName("params")]
        public new InitializeRequestParams? Params { get; init; }

        public static InitializeRequest From(RpcMessage message)
        {
            return new()
            {
                JsonRpc = message.JsonRpc,
                Id = message!.Id,
                Method = message.Method,
                Params = message.Params?.Deserialize<InitializeRequestParams>()
            };
        }
    }

    public record InitializeRequestParams
    {
        [JsonPropertyName("capabilities")]
        public JsonElement? Capabilities { get; init; }

        [JsonPropertyName("rootPath")]
        public string? RootPath { get; init; }

        [JsonPropertyName("processId")]
        public int? ProcessId { get; init; }
    }

    public record CompletionOptions
    {
        [JsonPropertyName("resolveProvider")]
        public bool? ResolveProvider { get; init; }

        [JsonPropertyName("triggerCharacters")]
        public char[]? TriggerCharacters { get; init; }
    }

    public record ServerCapabilities
    {
        [JsonPropertyName("completionProvider")]
        public CompletionOptions? CompletionProvider { get; init; }
    }

    public record InitializeResultParams
    {
        [JsonPropertyName("capabilities")]
        public ServerCapabilities? Capabilities { get; init; }
    }
}
